### ISSUE TITLE
False positives for -Wunused-but-set-variable with op assignment operators (e.g. `|=`)

### ISSUE THREAD

**Issue by luismarques** [llvm/llvm-project#51191]
|  |  |
| --- | --- |
| Bugzilla Link | [51849](https://llvm.org/bz51849) |
| Version | trunk |
| OS | All |
| CC | @dwblaikie,@zygoloid |

## Extended Description 
I think this is a false positive.
This happens with trunk (9aeecdfa8e91) and the latest RC (13.0.0-rc3):

$ cat test.c
void f(void) {
  int result = 0;
#ifdef BUG
    result |= 42;
#else
    result = result | 42;
#endif
}
$ ./bin/clang -Wunused-but-set-variable -c test.c
$ ./bin/clang -Wunused-but-set-variable -c test.c -DBUG
test.c:2:7: warning: variable 'result' set but not used [-Wunused-but-set-variable]
  int result = 0;
      ^
1 warning generated.

**Comment by Quuxplusone** [comment]
> GCC doesn't warn in this sort of case, but I think it's reasonable to warn

Agreed. The test case here is
```
int g(void);
void f(void) {
  int result = 0;
  for (int i = 0; i < 3; ++i)
    result |= g();
}
```
which is a very clear-cut case where `result` is written (multiple times), but all those writes are dead, because the variable is never read. This is the raison d'etre of `-Wunused-but-set-variable`. If you want the compiler to stop warning about the dead writes, then just remember to _use_ the value you've computed:
```
int f(void) {
  int result = 0;
  for (int i = 0; i < 3; ++i)
    result |= g();
  return result;  // FIX #1
}
```
or alternatively, suppress the warning via the cast-to-void idiom or (in C++) `[[maybe_unused]]`:
```
void f(void) {
  [[maybe_unused]] int result = 0;  // FIX #2
  (void)result;  // FIX #3
  for (int i = 0; i < 3; ++i)
    result |= g();
}
```
or of course compile with `-Wno-unused-but-set-variable`.
